clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -384,6 +389,32 @@ int attempts() {
         return attempts.get();
     }
 
+    /*
+     * Returns whether the leader Node has changed since the last attempt.
+     * @param node The Node currently thought of as the leader, which might be null.
+     * @return true if the leader has changed, otherwise false
+     */
+    boolean hasLeaderChanged(Node latestLeader) {
+        boolean leaderChanged = false;
+        if (latestLeader != null) {
+            // If we don't know the leader yet, we have not yet attempted to send to the leader
+            if (currentLeader == null) {
+                currentLeader = latestLeader;
+            } else {
+                // If the leader's node id has changed, this counts as a leader change
+                if (currentLeader.id() != latestLeader.id()) {
```

Comment:
We probably need to take leaderEpoch into consideration. For example, if the leader epoch has changed, even if the leader node remains the same, it's still worth to retry the batch now.

Reply:
It's going to take a bit of refactoring to get this optimisation to use the epoch cleanly. Rather than bloating this PR, I'll revert this part and then follow this PR with another with the epoch-based idea.

Reply:
Did we do this in the end?

Reply:
Yes, I believe so. If you look at the current `RecordAccumulator.shouldBackoff`, it knows when the leader changed and changes behaviour accordingly.

Reply:
The change was done along with KIP-951 changes.